databroker/broker.py

```diff
@@ -884,3 +902,38 @@ def _munge_time(t, timezone):
     """
     t = datetime.fromtimestamp(t)
     return timezone.localize(t).replace(microsecond=0).isoformat()
+
+
+class HeaderSourceShim(object):
+    '''Shim class to turn a mds object into a HeaderSource
+
+    This will presumably be deleted if this API makes it's way back down
+    into the implementations
+    '''
+    def __init__(self, mds):
+        self.mds = mds
+
+    def __call__(self, text_search=None, filters=None, **kwargs):
+        if filters is None:
+            filters = []
+        if text_search is not None:
+            query = {'$and': [{'$text': {'$search': text_search}}] + filters}
+        else:
+            # Include empty {} here so that '$and' gets at least one query.
+            _format_time(kwargs, self.mds.config['timezone'])
+            query = {'$and': [{}] + [kwargs] + filters}
+
+        starts = tuple(self.mds.find_run_starts(**query))
+
+        stops = tuple(_safe_get_stop(self.mds, s) for s in starts)
+        return zip(starts, stops)
+
+    def __getitem__(self, k):
+        return search(k, self)
+
+
+def _safe_get_stop(mds, s):
```

Comment:
If you hit this _before_ stop comes out or things went sideways so badly it did not come out.